src/prefect/utilities/local_cache.py

```diff
@@ -0,0 +1,195 @@
+    def __enter__(self) -> None:
+        self.lock.__enter__()
+
+    def __exit__(self, *args: Any) -> None:
+        self.lock.__exit__(*args)
+
+    def __reduce__(self) -> tuple:
+        return _SerializableLock, (self.reentrant, self.token)
+
+    def __repr__(self) -> str:
+        return "<%s: %s>" % (self.__class__.__name__, self.token)
+
+
+@atexit.register
+def clear_all_local_caches() -> None:
+    """On process exit, ensure all local caches are cleared.
+
+    For most executors, this should be a no-op (the cleanup should have already run).
+    """
+    for flow_run_id in list(_CACHES):
+        clear_local_cache(flow_run_id)
+
+
+def _init_cache() -> dict:
+    out = defaultdict(dict)  # type: dict
+    out["generators"] = []
+    return out
+
+
+_CACHES = defaultdict(_init_cache)  # type: dict
+
+
+def get_local_cache() -> dict:
+    # If not running inside a flow run, use a global cache This shouldn't
+    # happen in normal flow usage, but allows shared_object functions to be
+    # called from outside a flow run (for e.g. testing)
+    key = prefect.context.get("flow_run_id", "GLOBAL")
+    return _CACHES[key]
+
+
+def clear_local_cache(flow_run_id: str) -> None:
+    cache = _CACHES.pop(flow_run_id, None)
+    if cache is None:
+        return
+    logger = get_logger()
+    for generator in cache["generators"]:
+        try:
+            next(generator)
+        except StopIteration:
+            pass
+        except Exception:
+            logger.warning(
+                "Error while cleaning up `shared_object` function", exc_info=True
+            )
+
+
+def _default_key(*args: Any, **kwargs: Any) -> Any:
+    return (args, frozenset(kwargs.items()) if kwargs else None)
+
+
+@curry
+def shared_object(func: Callable, *, key: Callable = None) -> Callable:
+    """Marks the result of this helper function as a shared object to be reused
+    by other tasks.
+
+    Functions decorated with ``shared_object`` have their results stored in a
+    temporary cache specific to a flow run & process. If the function is called
+    with the same parameters during the same flow run in the same process, it
+    will not be re-evaluated. Upon completion of the flow run, the cache will
+    be cleared.
+
+    Note that ``shared_object`` functions should never be tasks themselves, but
+    should be called as helper-functions from inside tasks. They can be useful
+    for creating and reusing resources per-worker-process while avoiding
+    treating these clients as the results of tasks (which must be picklable
+    when running with multiple processes).
+
+    Args:
+        - func (Callable): the function to cache. May either `return` a result,
+            or `yield` one. If `yield` is used, the code block after the
+            `yield` statement will be run as a cleanup step upon completion of
+            the flow run (only one `yield` statement is permitted).
+        - key (Callable, optional): a function to use to generate the cache
```

Comment:
Might make more sense as `key_fn`, `keygen`, or `key_generator` ? As named the purpose of this variable and the `_default_key` function was initially confusing.